### ISSUE TITLE
is the cmd in launcher and worker pod different?

### ISSUE THREAD

**Issue by tingweiwu** [kubeflow/mpi-operator#58]
when I start a mpijob as follow
```
apiVersion: kubeflow.org/v1alpha1
kind: MPIJob
metadata:
  name: mpijob-horovod
spec:
  replicas: 2
  backoffLimit: 0
  template:
    spec:
      containers:
      - image: uber/horovod:0.13.10-tf1.9.0-torch0.4.0-py3.5
        args:
         - cp -rf /code /temp && cd /temp && mpirun python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod
        command:
         - /bin/bash
         - -c
        imagePullPolicy: IfNotPresent
        name: mpijob-horovod
        resources:
          limits:
            nvidia.com/gpu: 8
        volumeMounts:
        - mountPath: /code
          name: codevolume
          readOnly: true
      restartPolicy: Never
      volumes:
      - hostPath:
          path: /root/Horovod/local/tf_cnn_benchmarks
          type: DirectoryOrCreate
        name: codevolume
```

 I get an error `python: can't open file 'tf_cnn_benchmarks.py': [Errno 2] No such file or directory`.
 I check that the ` /temp` dir is existed in launcher  pod but not in the worker.
I guess the cmd `cp -rf /code /temp && cd /temp` have not executed in the worker pod.

when I change cmd into `cd /code && mpirun python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod`.  It works well.

so, is the cmd  in launcher and worker pod different?

thanks a lot if you reply

**Comment by rongou** [comment]
Yes that's currently a limitation. We use `PodSpec` for both the workers and the launcher, but they run different commands. The workers simply run `sleep 365d` and ignores the command set in the spec. More details here: https://github.com/kubeflow/community/blob/master/proposals/mpi-operator-proposal.md

Going forward we might want to separate them out, see #54.

**Comment by tingweiwu** [comment]
yes. I know the entrypoint cmd in launcher and workers pod are different as I describe pod when MPIJob running.  
I means that when podSpec cmd in MPIJob is `cd /code && mpirun python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod`

does `cd /code` execute in worker pod ? If not, how the workers find  `tf_cnn_benchmarks.py`?

maybe I am confused about mpirun communicates among  launcher pod and worker pods.

**Comment by rongou** [comment]
No the worker pods only execute `sleep 365d`. Are you modifying `tf_cnn_benchmarks.py` locally? If not, you can just put it in the container image. See https://github.com/kubeflow/mpi-operator/blob/master/examples/tensorflow-benchmarks/Dockerfile#L5.

**Comment by tingweiwu** [comment]
I can't put my `tf_cnn_benchmarks.py` in the container image. because I will run others script . the `tf_cnn_benchmarks.py` is just a example.
As my code is mount into pod.
when the entrypoint in `MPIJob Spec` is
```
cd /code && mpirun python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod. 
```
it works well.

I wonder as the `cd /code` not execute in worker pods. how the workers pod find the script  `tf_cnn_benchmarks.py`

and why I change entryponit in `MPIJob Spec` to
```
cp -rf /code /temp && cd /temp && mpirun python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod
```
it failed with the error ` python: can't open file 'tf_cnn_benchmarks.py': [Errno 2] No such file or directory`

**Comment by rongou** [comment]
I guess it depends on how your container image is built. If the `WORKDIR` is set to `/code`, then `cd /code` is basically a no-op.

**Comment by tingweiwu** [comment]
@rongou I have already checked the WORKDIR in container image early . 
It is the horovod examples dir. and when I haven't mount the `/root/Horovod/local/tf_cnn_benchmarks` into the container  path `/code`. you can see there is no `/code `dir as the following
so I can be sure that the `code` is not the WORKDIR in container image.

this makes me confused.

```
root@c8f2d92d9f10:/examples# ll
total 80
drwxr-xr-x. 2 root root 4096 Aug 31 07:33 ./
drwxr-xr-x. 1 root root 4096 Dec 19 06:14 ../
-rw-r--r--. 1 root root 8201 Aug 31 07:33 keras_imagenet_resnet50.py
-rw-r--r--. 1 root root 3223 Aug 31 07:33 keras_mnist.py
-rw-r--r--. 1 root root 4985 Aug 31 07:33 keras_mnist_advanced.py
-rw-r--r--. 1 root root 9558 Aug 31 07:33 pytorch_imagenet_resnet50.py
-rw-r--r--. 1 root root 5703 Aug 31 07:33 pytorch_mnist.py
-rw-r--r--. 1 root root 5039 Aug 31 07:33 tensorflow_mnist.py
-rw-r--r--. 1 root root 7296 Aug 31 07:33 tensorflow_mnist_estimator.py
-rw-r--r--. 1 root root 9723 Aug 31 07:33 tensorflow_word2vec.py
root@c8f2d92d9f10:/examples# ll /code
ls: cannot access '/code': No such file or directory
```

**Comment by rongou** [comment]
Hmm not sure what's going on. Where are you seeing the "no such file" error? Is it in the launcher pod or the worker pod?

**Comment by tingweiwu** [comment]
actually I am not sure "no such file" error is from launcher pod or the worker pod. firstly, I find that the ` /temp`   dir exsit  in launcher pod. and not in the worker pod 

cmd is 
```
cp -rf /code /temp && cd /temp && mpirun python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod
```
the error logs is as following, and launcher pod goes into Error  status
```
# kubectl logs mpijob-0-launcher-hmtgn
+ POD_NAME=mpijob-0-worker-0
+ shift
+ /opt/kube/kubectl exec mpijob-0-worker-0 -- /bin/sh -c     PATH=/usr/local/bin:$PATH ; export PATH ; LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH ; export LD_LIBRARY_PATH ; DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH ; export DYLD_LIBRARY_PATH ;   /usr/local/bin/orted -mca ess "env" -mca ess_base_jobid "2398027776" -mca ess_base_vpid 1 -mca ess_base_num_procs "3" -mca orte_node_regex "mpijob-0-launcher-hmtgn,mpijob-0-worker-0,mpijob-0-worker-1@0(3)" -mca orte_hnp_uri "2398027776.0;tcp://172.16.217.76:45095" -mca plm "rsh" -mca plm_rsh_agent "/etc/mpi/kubexec.sh" -mca orte_default_hostfile "/etc/mpi/hostfile" -mca pmix "^s1,s2,cray,isolated"
+ POD_NAME=mpijob-0-worker-1
+ shift
+ /opt/kube/kubectl exec mpijob-0-worker-1 -- /bin/sh -c     PATH=/usr/local/bin:$PATH ; export PATH ; LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH ; export LD_LIBRARY_PATH ; DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH ; export DYLD_LIBRARY_PATH ;   /usr/local/bin/orted -mca ess "env" -mca ess_base_jobid "2398027776" -mca ess_base_vpid 2 -mca ess_base_num_procs "3" -mca orte_node_regex "mpijob-0-launcher-hmtgn,mpijob-0-worker-0,mpijob-0-worker-1@0(3)" -mca orte_hnp_uri "2398027776.0;tcp://172.16.217.76:45095" -mca plm "rsh" -mca plm_rsh_agent "/etc/mpi/kubexec.sh" -mca orte_default_hostfile "/etc/mpi/hostfile" -mca pmix "^s1,s2,cray,isolated"
python: can't open file 'tf_cnn_benchmarks.py': [Errno 2] No such file or directory
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
-------------------------------------------------------
python: can't open file 'tf_cnn_benchmarks.py': [Errno 2] No such file or directory
--------------------------------------------------------------------------
mpirun.real detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[36591,1],1]
  Exit code:    2
--------------------------------------------------------------------------

```

when change cmd into 
```
 cd /code && mpirun python tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod
```
the logs as following
```
# kubectl logs mpijob-0-launcher-4p7zt
+ POD_NAME=mpijob-0-worker-0
+ shift
+ /opt/kube/kubectl exec mpijob-0-worker-0 -- /bin/sh -c     PATH=/usr/local/bin:$PATH ; export PATH ; LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH ; export LD_LIBRARY_PATH ; DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH ; export DYLD_LIBRARY_PATH ;   /usr/local/bin/orted -mca ess "env" -mca ess_base_jobid "1473839104" -mca ess_base_vpid 1 -mca ess_base_num_procs "3" -mca orte_node_regex "mpijob-0-launcher-4p7zt,mpijob-0-worker-0,mpijob-0-worker-1@0(3)" -mca orte_hnp_uri "1473839104.0;tcp://172.16.217.78:46472" -mca plm "rsh" -mca plm_rsh_agent "/etc/mpi/kubexec.sh" -mca orte_default_hostfile "/etc/mpi/hostfile" -mca pmix "^s1,s2,cray,isolated"
+ POD_NAME=mpijob-0-worker-1
+ shift
+ /opt/kube/kubectl exec mpijob-0-worker-1 -- /bin/sh -c     PATH=/usr/local/bin:$PATH ; export PATH ; LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH ; export LD_LIBRARY_PATH ; DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH ; export DYLD_LIBRARY_PATH ;   /usr/local/bin/orted -mca ess "env" -mca ess_base_jobid "1473839104" -mca ess_base_vpid 2 -mca ess_base_num_procs "3" -mca orte_node_regex "mpijob-0-launcher-4p7zt,mpijob-0-worker-0,mpijob-0-worker-1@0(3)" -mca orte_hnp_uri "1473839104.0;tcp://172.16.217.78:46472" -mca plm "rsh" -mca plm_rsh_agent "/etc/mpi/kubexec.sh" -mca orte_default_hostfile "/etc/mpi/hostfile" -mca pmix "^s1,s2,cray,isolated"
TensorFlow:  1.10
Model:       resnet101
Dataset:     imagenet (synthetic)
Mode:        training
SingleSess:  False
Batch size:  128 global
             64.0 per device
Num batches: 10000
Num epochs:  1.00
Devices:     ['horovod/gpu:0', 'horovod/gpu:1']
Data format: NCHW
Optimizer:   sgd
Variables:   horovod
==========
TensorFlow:  1.10

.....................................
```

**Comment by rongou** [comment]
OK I think I know what's going on. When `mpirun` sets up the communication, it actually passes the working directory to all the workers. Since you have `/code` mounted on all the pods, it works if you use that directory. But since the worker pods only run `sleep 365d`, they wouldn't have the `/temp` directory, thus failing to locate the python file.

There is not much we can do right now until we fix #54.

**Comment by tingweiwu** [comment]
well, is there any play to fix #54?
Indeed I've noticed your invitation on fixing this problem on issus #53. I wish you don't mind that I haven't done it yet cause I'm totally new to GO language.

**Comment by rongou** [comment]
We have some work planned to enhance the mpi operator, we can put #54 in the list. Closing this for now.
